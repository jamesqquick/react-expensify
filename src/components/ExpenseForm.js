import 'react-dates/initialize';
import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            amount: props.expense
                ? (props.expense.amount / 100).toString()
                : '0',
            note: props.expense ? props.expense.note : '',
            createdAt: moment(),
            calendarFocused: false,
            error: ''
        };
    }

    onDescriptionChange = e => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    };

    onNoteChange = e => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    };

    onAmountChange = e => {
        const amount = e.target.value;
        //start with - caret
        // \d digit, * for any amount
        //optional stuff in parenthesis followed by question mark
        //ends with - dollar sign
        const regex = /^\d{1,}(\.\d{0,2})?$/;
        if (!amount || amount.match(regex)) {
            this.setState(() => ({ amount }));
        }
    };

    onDateChange = createdAt => {
        //dont allow to clear date
        if (createdAt) {
            this.setState(() => ({ createdAt }));
        }
    };

    onDateFocusChanged = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }));
    };

    onFormSubmit = e => {
        e.preventDefault();
        //Check description and amount
        if (!this.state.description || !this.state.amount) {
            //Set error
            this.setState(() => ({
                error: 'Make sure to fill in description and amount.'
            }));
        } else {
            //Clear error
            this.setState(() => ({
                error: ''
            }));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            });
        }
    };

    render() {
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onFormSubmit}>
                    <input
                        type="text"
                        placeholder="Description"
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    />
                    <input
                        type="text"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onDateFocusChanged}
                        numberOfMonths={1}
                        isOutsideRange={day => false}
                    />
                    <textarea
                        placeholder="Add a note for expense"
                        onChange={this.onNoteChange}
                    />
                    <button>Add Expense</button>
                </form>
            </div>
        );
    }
}

export default ExpenseForm;
