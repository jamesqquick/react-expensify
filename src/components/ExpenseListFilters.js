import React from 'react';
import { connect } from 'react-redux';
import {
    setTextFilter,
    sortByDate,
    sortByAmount,
    setStartDate,
    setEndDate
} from '../actions/filters';

import { DateRangePicker } from 'react-dates';

export class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null
    };

    onTextChange = e => {
        this.props.setTextFilter(e.target.value);
    };

    onSortChange = e => {
        if (e.target.value === 'date') {
            this.props.sortByDate();
        } else if (e.target.value === 'amount') {
            this.props.sortByAmount();
        }
    };

    render() {
        return (
            <div>
                <input
                    type="text"
                    id="textInput"
                    value={this.props.filters.text}
                    onChange={this.onTextChange}
                />
                <select
                    value={this.props.filters.filterBy}
                    onChange={this.onSortChange}
                >
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onDateRangeFocusChanged}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                    showClearDates={true}
                    startDateId="startDate"
                    endDateId="endDate"
                />
            </div>
        );
    }
    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };

    onDateRangeFocusChanged = calendarFocused => {
        this.setState(() => ({ calendarFocused }));
    };
}

const mapStateToProps = state => ({ filters: state.filters });

const mapDispatchToProps = dispatch => {
    return {
        setTextFilter: text => dispatch(setTextFilter(text)),
        sortByDate: () => dispatch(sortByDate()),
        sortByAmount: () => dispatch(sortByAmount()),
        setStartDate: startDate => dispatch(setStartDate(startDate)),
        setEndDate: endDate => dispatch(setEndDate(endDate))
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
