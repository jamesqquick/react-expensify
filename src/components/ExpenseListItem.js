import React from 'react';
import { deleteExpense } from '../actions/expenses';
import { NavLink } from 'react-router-dom';

const ExpenseListItem = ({ id, description, amount, createdAt }) => {
	return (
		<div>
			<h3>{description}</h3>
			<p>
				Amount: {amount} - Created At: {createdAt}
			</p>
			<NavLink to={`/edit/${id}`}>
				<button>Edit</button>
			</NavLink>
		</div>
	);
};

export default ExpenseListItem;
