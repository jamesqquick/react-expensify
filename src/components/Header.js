import React from 'react';
import { NavLink } from 'react-router-dom';
import { startLogout } from '../actions/auth';
import { connect } from 'react-redux';

export const HeaderComponent = ({ startLogout }) => (
    <header>
        <h1>Expensify</h1>
        <NavLink to="/dashboard" activeClassName="active">
            Dashboard
        </NavLink>
        <NavLink to="/create" activeClassName="active">
            Create Expense
        </NavLink>
        <NavLink to="/help" activeClassName="active">
            Help
        </NavLink>
        <button onClick={startLogout} id="logoutBtn">
            Logout
        </button>
    </header>
);

const mapDispatchToProps = dispatch => ({
    startLogout: () => dispatch(startLogout())
});
export default connect(undefined, mapDispatchToProps)(HeaderComponent);
