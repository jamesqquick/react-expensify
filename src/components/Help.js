import React from 'react';

const HelpComponent = () => <div>Help</div>;

export default HelpComponent;
