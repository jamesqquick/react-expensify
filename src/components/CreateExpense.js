import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { startAddExpense } from '../actions/expenses';

export class CreateExpense extends React.Component {
    onSubmit = expense => {
        this.props.startAddExpense(expense);
        this.props.history.push('/');
    };

    render() {
        return (
            <div>
                <p>Add</p>
                <ExpenseForm onSubmit={this.onSubmit} />
            </div>
        );
    }
}

const mapDispatchToProps = dispatch => {
    return {
        startAddExpense: expense => dispatch(startAddExpense(expense))
    };
};

export default connect(undefined, mapDispatchToProps)(CreateExpense);
