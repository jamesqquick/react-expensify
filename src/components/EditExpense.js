import React from 'react';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startDeleteExpense } from '../actions/expenses';
import { connect } from 'react-redux';

export class EditExpense extends React.Component {
    render() {
        return (
            <div>
                Edit expense: {this.props.match.params.id}
                <ExpenseForm
                    onSubmit={this.submitEdit}
                    expense={this.props.expense}
                />
                <button
                    id="startDeleteExpenseBtn"
                    onClick={this.startDeleteExpense}
                >
                    Delete
                </button>
            </div>
        );
    }

    submitEdit = expense => {
        this.props.startEditExpense(this.props.expense.id, expense);
        this.props.history.push('/');
    };

    startDeleteExpense = () => {
        this.props.startDeleteExpense(this.props.expense.id);
        this.props.history.push('/');
    };
}

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find(
            expense => expense.id === props.match.params.id
        )
    };
};

const mapDispatchToProps = dispatch => {
    return {
        startEditExpense: (id, expense) =>
            dispatch(startEditExpense(id, expense)),
        startDeleteExpense: data => dispatch(startDeleteExpense(data))
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(EditExpense);
