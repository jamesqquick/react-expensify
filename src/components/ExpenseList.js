import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

//Basic component
export const ExpenseList = props => (
	<div>
		<h1>Expense List</h1>
		{props.expenses.length === 0 && <p>No Expenses</p>}
		{props.expenses.map(expense => {
			return <ExpenseListItem key={expense.id} {...expense} />;
		})}
	</div>
);

//mapping of state to props
const mapStateToProps = state => ({
	expenses: selectExpenses(state.expenses, state.filters)
});

//exported component with props
export default connect(mapStateToProps)(ExpenseList);
