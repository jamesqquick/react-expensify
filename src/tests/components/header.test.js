import React from 'react';
import { HeaderComponent } from '../../components/Header';
import { shallow } from 'enzyme';

test('should render header correctly', () => {
    const wrapper = new shallow(<HeaderComponent startLogout={() => {}} />);
    expect(wrapper).toMatchSnapshot();
});

test('should call startLogout on button click', () => {
    const startLogout = jest.fn();
    const wrapper = new shallow(<HeaderComponent startLogout={startLogout} />);
    wrapper.find('#logoutBtn').simulate('click');
    expect(startLogout).toHaveBeenCalled();
});
