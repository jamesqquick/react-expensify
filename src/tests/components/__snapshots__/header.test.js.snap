// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render header correctly 1`] = `
<header>
  <h1>
    Expensify
  </h1>
  <NavLink
    activeClassName="active"
    ariaCurrent="true"
    to="/dashboard"
  >
    Dashboard
  </NavLink>
  <NavLink
    activeClassName="active"
    ariaCurrent="true"
    to="/create"
  >
    Create Expense
  </NavLink>
  <NavLink
    activeClassName="active"
    ariaCurrent="true"
    to="/help"
  >
    Help
  </NavLink>
  <button
    id="logoutBtn"
    onClick={[Function]}
  >
    Logout
  </button>
</header>
`;
